[VPlan] Simplify Plan's entry in removeBranchOnConst.

fhahn opened the pull request:
After https://github.com/llvm/llvm-project/pull/153643, there may be a
BranchOnCond with constant condition in the entry block.

Simplify those in removeBranchOnConst. This removes a number of
redundant conditional branch from entry blocks.

In some cases, it may also make the original scalar loop unreachable,
because we know it will never execute. In that case, we need to remove
the loop from LoopInfo, because all unreachable blocks may dominate each
other, making LoopInfo invalid. In those cases, we can also completely
remove the loop, for which I'll share a follow-up patch.

A couple of places that assume the scalar loop to still be valid need
updating.

Depends on https://github.com/llvm/llvm-project/pull/153643.

fhahn committed:
[VPlan] Simplify Plan's entry in removeBranchOnConst.

fhahn committed:
[LV] Add variable trip count for test.

fhahn reviewed (commented):
ping :)

ayalz commented on llvm/test/Transforms/LoopVectorize/make-followup-loop-id.ll:
```diff
@@ -67,53 +67,17 @@ define void @f(ptr noundef captures(none) %a, float noundef %x) {
-; CHECK-NEXT:    br label %[[EXIT]]
+; CHECK-NEXT:    [[IV_NEXT]] = add nuw nsw i64 [[IV]], 1
+; CHECK-NEXT:    [[COMP:%.*]] = icmp eq i64 [[IV_NEXT]], 1024
+; CHECK-NEXT:    br i1 [[COMP]], label %[[EXIT]], label %[[FOR_BODY]], !llvm.loop [[LOOP3:![0-9]+]]
```
Dropped unrolling (requested, of vector loop) of unreachable scalar loop?

ayalz commented on llvm/test/Transforms/LoopVectorize/min-trip-count-known-via-scev.ll:
```diff
@@ -288,39 +287,14 @@ define void @loop_never_executes_precondition_order_2_predicates_flipped(i64 %st
 ; CHECK-NEXT:    br i1 [[PRE_0]], label %[[LOOP_PREHEADER:.*]], label %[[EXIT]]
 ; CHECK:       [[LOOP_PREHEADER]]:
-; CHECK-NEXT:    [[TMP0:%.*]] = sub i64 1, [[START]]
-; CHECK-NEXT:    br i1 true, label %[[SCALAR_PH:.*]], label %[[VECTOR_PH:.*]]
```
This skips the vector loop! A case where "loop_never_executes_..."

ayalz commented on llvm/test/Transforms/LoopVectorize/no_outside_user.ll:
```diff
@@ -23,7 +23,7 @@ define i32 @test1()  {
 ; CHECK-NEXT:    [[TMP1:%.*]] = sub i32 [[SMAX]], [[B_PROMOTED]]
 ; CHECK-NEXT:    [[MIN_ITERS_CHECK:%.*]] = icmp ult i32 [[TMP1]], 2
-; CHECK-NEXT:    br i1 [[MIN_ITERS_CHECK]], label %[[SCALAR_PH:.*]], label %[[VECTOR_PH:.*]]
+; CHECK-NEXT:    br i1 [[MIN_ITERS_CHECK]], label %[[_LR_PH_I:.*]], label %[[VECTOR_PH:.*]]
```
Changing of label regex's a bit confusing.

ayalz commented on llvm/test/Transforms/LoopVectorize/single_early_exit_with_outer_loop.ll:
```diff
@@ -1,12 +1,11 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
```
Do these assertions need to be autogenerated here now. Doing so seems to only be dropping checks.

ayalz commented on llvm/test/Transforms/LoopVectorize/trunc-reductions.ll:
```diff
@@ -110,16 +108,15 @@ define i16 @reduction_xor_trunc(ptr noalias nocapture %ptr) {
+; CHECK-NEXT:    br i1 poison, label [[FOR_END]], label [[FOR_BODY]]
 ; CHECK:       for.end:
-; CHECK-NEXT:    [[XOR_LCSSA_OFF0:%.*]] = phi i16 [ poison, [[FOR_BODY]] ], [ [[TMP4]], [[MIDDLE_BLOCK]] ]
+; CHECK-NEXT:    [[XOR_LCSSA_OFF0:%.*]] = call i16 @llvm.vector.reduce.xor.v8i16(<8 x i16> [[TMP2]])
```
Hmm, vector.reduce sunk from middle to exit, once redundant edge from  scalar loop to exit is gone.

ayalz commented on llvm/test/Transforms/LoopVectorize/vector-loop-backedge-elimination.ll:
```diff
@@ -53,76 +53,35 @@ define void @test_tc_less_than_16(ptr %A, i64 %N) {
 ; VF8UF2-SAME: ptr [[A:%.*]], i64 [[N:%.*]]) {
 ; VF8UF2-NEXT:  [[ENTRY:.*]]:
 ; VF8UF2-NEXT:    [[AND:%.*]] = and i64 [[N]], 15
-; VF8UF2-NEXT:    br i1 true, label %[[SCALAR_PH:.*]], label %[[VECTOR_PH:.*]]
```
Another case skipping the vector loop.

ayalz commented on llvm/test/Transforms/LoopVectorize/version-stride-with-integer-casts.ll:
```diff
@@ -468,37 +467,14 @@ define void @sext_of_i1_stride(i1 %g, ptr %dst) mustprogress {
-; CHECK-NEXT:    [[TMP0:%.*]] = add i64 [[UMAX]], -1
-; CHECK-NEXT:    [[TMP1:%.*]] = udiv i64 [[TMP0]], [[G_64]]
-; CHECK-NEXT:    [[TMP2:%.*]] = add nuw nsw i64 [[TMP1]], 1
-; CHECK-NEXT:    br i1 true, label [[SCALAR_PH:%.*]], label [[VECTOR_SCEVCHECK:%.*]]
```
ditto

ayalz commented on llvm/test/Transforms/LoopVectorize/vplan-printing-before-execute.ll:
```diff
@@ -12,15 +12,13 @@ define void @test_tc_less_than_16(ptr %A, i64 %N) {
 ; CHECK-NEXT: Live-in vp<[[VFxUF:%.+]]> = VF * UF
 ; CHECK-NEXT: Live-in vp<[[VTC:%.+]]> = vector-trip-count
-; CHECK-NEXT: vp<[[TC:%.+]]> = original trip-count
+; CHECK-NEXT: ir<16> = original trip-count
```
Simplification due to branch-on-const of entry?

ayalz commented on llvm/test/Transforms/LoopVectorize/vplan-printing-before-execute.ll:
```diff
@@ -88,31 +80,17 @@ define void @test_tc_less_than_16(ptr %A, i64 %N) {
 ; CHECK-NEXT: }
 ;
 entry:
-  %and = and i64 %N, 15
```
Ah, test changed?

ayalz commented on llvm/test/Transforms/PhaseOrdering/X86/vdiv-nounroll.ll:
```diff
@@ -20,14 +20,32 @@ define void @vdiv(ptr %a, float %b) #0 {
 ; CHECK-NEXT:    [[BROADCAST_SPLAT:%.*]] = shufflevector <4 x float> [[BROADCAST_SPLATINSERT]], <4 x float> poison, <4 x i32> zeroinitializer
 ; CHECK-NEXT:    [[TMP0:%.*]] = fdiv fast <4 x float> splat (float 1.000000e+00), [[BROADCAST_SPLAT]]
+; CHECK-NEXT:    [[TMP4:%.*]] = fdiv fast <4 x float> splat (float 1.000000e+00), [[BROADCAST_SPLAT]]
+; CHECK-NEXT:    [[TMP2:%.*]] = fdiv fast <4 x float> splat (float 1.000000e+00), [[BROADCAST_SPLAT]]
+; CHECK-NEXT:    [[TMP15:%.*]] = fdiv fast <4 x float> splat (float 1.000000e+00), [[BROADCAST_SPLAT]]
 ; CHECK-NEXT:    br label [[VECTOR_BODY:%.*]]
 ; CHECK:       vector.body:
 ; CHECK-NEXT:    [[INDEX:%.*]] = phi i64 [ 0, [[ENTRY:%.*]] ], [ [[INDEX_NEXT:%.*]], [[VECTOR_BODY]] ]
 ; CHECK-NEXT:    [[TMP1:%.*]] = getelementptr inbounds nuw float, ptr [[A:%.*]], i64 [[INDEX]]
 ; CHECK-NEXT:    [[WIDE_LOAD:%.*]] = load <4 x float>, ptr [[TMP1]], align 4, !tbaa [[TBAA3:![0-9]+]]
 ; CHECK-NEXT:    [[TMP3:%.*]] = fmul fast <4 x float> [[WIDE_LOAD]], [[TMP0]]
 ; CHECK-NEXT:    store <4 x float> [[TMP3]], ptr [[TMP1]], align 4, !tbaa [[TBAA3]]
-; CHECK-NEXT:    [[INDEX_NEXT]] = add nuw i64 [[INDEX]], 4
+; CHECK-NEXT:    [[TMP6:%.*]] = getelementptr inbounds nuw float, ptr [[A]], i64 [[INDEX]]
+; CHECK-NEXT:    [[TMP7:%.*]] = getelementptr inbounds nuw i8, ptr [[TMP6]], i64 16
+; CHECK-NEXT:    [[WIDE_LOAD_1:%.*]] = load <4 x float>, ptr [[TMP7]], align 4, !tbaa [[TBAA3]]
+; CHECK-NEXT:    [[TMP8:%.*]] = fmul fast <4 x float> [[WIDE_LOAD_1]], [[TMP4]]
+; CHECK-NEXT:    store <4 x float> [[TMP8]], ptr [[TMP7]], align 4, !tbaa [[TBAA3]]
+; CHECK-NEXT:    [[TMP9:%.*]] = getelementptr inbounds nuw float, ptr [[A]], i64 [[INDEX]]
+; CHECK-NEXT:    [[TMP10:%.*]] = getelementptr inbounds nuw i8, ptr [[TMP9]], i64 32
+; CHECK-NEXT:    [[WIDE_LOAD_2:%.*]] = load <4 x float>, ptr [[TMP10]], align 4, !tbaa [[TBAA3]]
+; CHECK-NEXT:    [[TMP11:%.*]] = fmul fast <4 x float> [[WIDE_LOAD_2]], [[TMP2]]
+; CHECK-NEXT:    store <4 x float> [[TMP11]], ptr [[TMP10]], align 4, !tbaa [[TBAA3]]
+; CHECK-NEXT:    [[TMP12:%.*]] = getelementptr inbounds nuw float, ptr [[A]], i64 [[INDEX]]
+; CHECK-NEXT:    [[TMP13:%.*]] = getelementptr inbounds nuw i8, ptr [[TMP12]], i64 48
+; CHECK-NEXT:    [[WIDE_LOAD_3:%.*]] = load <4 x float>, ptr [[TMP13]], align 4, !tbaa [[TBAA3]]
+; CHECK-NEXT:    [[TMP14:%.*]] = fmul fast <4 x float> [[WIDE_LOAD_3]], [[TMP15]]
+; CHECK-NEXT:    store <4 x float> [[TMP14]], ptr [[TMP13]], align 4, !tbaa [[TBAA3]]
+; CHECK-NEXT:    [[INDEX_NEXT]] = add nuw nsw i64 [[INDEX]], 16
```
?

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-simplify-branch-on-const-entry

fhahn committed:
!fixup address latest comments, thanks!

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-simplify-branch-on-const-entry

fhahn committed:
!fixup address latest comments,t hanks!

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-simplify-branch-on-const-entry

fhahn committed:
!Fxiup address comments, thanks

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-simplify-branch-on-const-entry

fhahn committed:
!fixup fix profile info

fhahn commented on llvm/test/Transforms/LoopVectorize/check-prof-info.ll:
```diff
@@ -13,33 +13,33 @@ target datalayout = "e-m:e-i64:64-f80:128-n8:16:32:64-S128"
 ; CHECK:    br label [[VECTOR_BODY:%.*]]
 ; CHECK:  vector.body:
-; CHECK:    br i1 [[TMP6:%.*]], label [[MIDDLE_BLOCK:%.*]], label [[VECTOR_BODY]], !prof [[PROF1:![0-9]+]], !llvm.loop [[LOOP2:![0-9]+]]
+; CHECK:    br i1 [[TMP4:%.*]], label [[MIDDLE_BLOCK:%.*]], label [[VECTOR_BODY]], !llvm.loop [[LOOP0:![0-9]+]]
```
The original scalar loop isn't reachable any longer (scalar.ph doesn't have a predecessor), but the vector loop should still maintain the profile info.

I fixed the code, it requires retrieving the average trip count before the scalar loop has been destroyed, and updating either the vector or remainder loop profile info, depending on whether they exist.

fhahn commented on llvm/test/Transforms/LoopVectorize/X86/pr81872.ll:
```diff
@@ -37,16 +37,16 @@ define void @test(ptr noundef align 8 dereferenceable_or_null(16) %arr) #0 {
 ; CHECK-NEXT:    [[VEC_IND_NEXT]] = add <4 x i64> [[VEC_IND]], splat (i64 -4)
 ; CHECK-NEXT:    [[TMP9:%.*]] = icmp eq i64 [[INDEX_NEXT]], 12
-; CHECK-NEXT:    br i1 [[TMP9]], label [[MIDDLE_BLOCK:%.*]], label [[VECTOR_BODY]], !prof [[PROF1:![0-9]+]], !llvm.loop [[LOOP2:![0-9]+]]
+; CHECK-NEXT:    br i1 [[TMP9]], label [[MIDDLE_BLOCK:%.*]], label [[VECTOR_BODY]], !llvm.loop [[LOOP0:![0-9]+]]
```
Should be fixed now

fhahn committed:
!fixup fix formatting

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-simplify-branch-on-const-entry

fhahn committed:
!fixup update new tests

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-simplify-branch-on-const-entry

fhahn committed:
!fixup move out loop metadata update

fhahn committed:
!fixup move

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-simplify-branch-on-const-entry

fhahn committed:
!fixup move more

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-simplify-branch-on-const-entry

fhahn committed:
[VPlan] Consolidate logic to update loop metadata and profile info.

This patch consolidates updating loop metadata and profile info for both
the remainder and vector loops in a single place. This is NFC, modulo
consistently applying vectorization specific metadata also in the
experimental VPlan-native path.

Split off from https://github.com/llvm/llvm-project/pull/154510.

fhahn committed:
Merge branch 'main' into vplan-simplify-branch-on-const-entry-tmp

fhahn committed:
!fixu update on top of main

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-simplify-branch-on-const-entry-tmp

fhahn committed:
!fixup fix formatting

fhahn reviewed (commented):
The diff should be cleaned up now after consolidating updateing loop metadata and profile info in 8796dfdcbae0fbf6c58f80379c73f62f36a57653

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-simplify-branch-on-const-entry

fhahn committed:
!fixup update tests after merge

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-simplify-branch-on-const-entry

fhahn commented:
ping :)

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-simplify-branch-on-const-entry

fhahn committed:
!fixup address latest comments, thanks!

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/intrinsiccost.ll:
```diff
@@ -1,6 +1,6 @@
 ; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
-; RUN: opt -passes=loop-vectorize -S  %s | FileCheck %s --check-prefix=CHECK
-; RUN: opt -passes=loop-vectorize -debug-only=loop-vectorize -disable-output %s 2>&1 | FileCheck %s --check-prefix=CHECK-COST
+; RUN: opt -passes=loop-vectorize,instcombine,simplifycfg < %s -S -o - | FileCheck %s --check-prefix=CHECK
```
Yes, this was added back by accident when updating to latest main

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/optsize_minsize.ll:
```diff
@@ -40,14 +40,14 @@ define void @always_vectorize(ptr %p, i32 %x) {
 ; DEFAULT-NEXT:    [[INDVARS_IV_NEXT]] = add nuw nsw i64 [[INDVARS_IV]], 1
 ; DEFAULT-NEXT:    [[EXITCOND_NOT:%.*]] = icmp eq i64 [[INDVARS_IV_NEXT]], 4
-; DEFAULT-NEXT:    br i1 [[EXITCOND_NOT]], label %[[FOR_COND_CLEANUP]], label %[[FOR_BODY]], !llvm.loop [[LOOP0:![0-9]+]]
+; DEFAULT-NEXT:    br i1 [[EXITCOND_NOT]], label %[[FOR_COND_CLEANUP]], label %[[FOR_BODY]]
```
same as above, loop now unreachable

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/strict-fadd.ll:
```diff
@@ -970,7 +970,7 @@ loop:
   %for.next = sitofp i32 %iv to double
   %iv.next = add nsw i32 %iv, 1
-  %ec = icmp eq i32 %iv.next, 1025
+  %ec = icmp eq i32 %iv.next, %n
```
Without it there would be no merge phis in `scalar.ph`, which seems to be what the test needs to check.

fhahn committed:
!fixup restore tests

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/sve-epilog-vscale-fixed.ll:
```diff
@@ -145,7 +141,7 @@ for.body:
   store i8 1, ptr %arrayidx, align 1
   %iv.next = add nuw nsw i64 %iv, 1
-  %exitcond = icmp ne i64 %iv.next, %n
+  %exitcond = icmp ne i64 %iv.next, 1024
```
Restored the code, thanks

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -1723,7 +1739,21 @@ void LoopVectorizationPlanner::updateLoopMetadataAndProfileInfo(
   // iterations of the loop are handled in one vector iteration, so instead
   // use the value of vscale used for tuning.
-  setProfileInfoAfterUnrolling(OrigLoop, VectorLoop, OrigLoop, EstimatedVFxUF);
+  if (OrigAverageTripCount) {
```
Early exit, as originally in llvm::setProfileInfoAfterUnrolling()?

fhahn replied:
Done thanks!

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1927,7 +1927,7 @@ void VPlanTransforms::removeBranchOnConst(VPlan &Plan) {
            vp_depth_first_shallow(Plan.getEntry()))) {
     VPValue *Cond;
-    if (VPBB->getNumSuccessors() != 2 || VPBB == Plan.getEntry() ||
+    if (VPBB->getNumSuccessors() != 2 || VPBB->empty() ||
```
Thanks. Would it suffice to check only if VPBB terminator matches a BranchOnCond? Is the emptiness check needed to prevent failure of this match? Check for two successors could then be an assert.

ayalz replied:
Understood, plus even w/ a pair of successors VPBB may be empty. Is it then (necessary and) suffice to check
```
    if (VPBB->empty() ||
        !match(&VPBB->back(), m_BranchOnCond(m_VPValue(Cond))))
      continue;

    assert(VPBB->getNumSuccessors() == 2 && "Two successors expected for BranchOnCond"); 
```
but redundant first check for two-successors may early-continue faster?

fhahn replied:
Updated to the simpler check (together with simpler comment),thanks

ayalz reviewed (approved):
This LGTM with minor last suggestions. Pending @david-arm's consent.

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-simplify-branch-on-const-entry

fhahn committed:
!fixup address comments, thanks

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-simplify-branch-on-const-entry

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7191,6 +7194,19 @@ DenseMap<const SCEV *, Value *> LoopVectorizationPlanner::executePlan(
+      return OptimizationRemarkAnalysis(DEBUG_TYPE, "VectorizationDead",
+                                        OrigLoop->getStartLoc(),
+                                        OrigLoop->getHeader())
+             << "Created vector loop never executes due insufficient trip "
```
```suggestion
             << "Created vector loop never executes due to insufficient trip "
```

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7377,6 +7387,16 @@ DenseMap<const SCEV *, Value *> LoopVectorizationPlanner::executePlan(
   }
 
+  if (BestVPlan.getScalarPreheader()->getNumPredecessors() == 0) {
+    // If the original loop became unreachable, we need to delete it.
```
Having it in VPlan::execute is great!
As a potential follow-up: the VPIRBasicBlock of the scalar preheader and/or a VPIRRegionBlock of the scalar loop could be marked for destruction in VPlan if unreachable (as in CreatedBlocks), indicating that its loop should be removed from LoopInfo etc.

fhahn replied:
Yep, they are implicitly marked for destruction (all VPIRBasicBlocks that are unreachable are tracked in CreatedBlocks) and can be destroyed in the destructor; but they need LoopInfo/DT passed somehow, to notify those about the removal

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-simplify-branch-on-const-entry

fhahn committed:
!fixup adjust check and message.

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-simplify-branch-on-const-entry

fhahn committed:
Merge branch 'main' into vplan-simplify-branch-on-const-entry
